Make Future store weak pointers to storages

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60943**

In https://github.com/pytorch/pytorch/pull/60470 we made Future store Storages rather than store references to their DataPtrs (because these references could go stale...). However this meant that the Future could keep the Storage alive, and thus keep its memory allocated, even after the user was done with it. We fix it here by instead storing a weak ptr to that Storage (well, in fact to the StorageImpl, but it's the same).

Differential Revision: [D29454104](https://our.internmc.facebook.com/intern/diff/D29454104/)

lw committed:
Make Future store weak pointers to storages

In https://github.com/pytorch/pytorch/pull/60470 we made Future store Storages rather than store references to their DataPtrs (because these references could go stale...). However this meant that the Future could keep the Storage alive, and thus keep its memory allocated, even after the user was done with it. We fix it here by instead storing a weak ptr to that Storage (well, in fact to the StorageImpl, but it's the same).

Differential Revision: [D29454104](https://our.internmc.facebook.com/intern/diff/D29454104/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -658,11 +659,14 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
+    for (const WeakStorage& weak_storage : storages_) {
+      c10::intrusive_ptr<c10::StorageImpl> storage = weak_storage.lock();
+      if (!storage) {
+        continue;
```
do we need any warning here?

lw replied:
At this point I think if we did it would fire all the time. Hence for the moment I'm perhaps more inclined to keep this silent: this way we don't introduce any (visible) functional change wrt the previous code, the only difference being that Future doesn't cause any memory to "leak".

mrshenli reviewed (approved):
LGTM! Thanks!

lw committed:
Update on "Make Future store weak pointers to storages"


In https://github.com/pytorch/pytorch/pull/60470 we made Future store Storages rather than store references to their DataPtrs (because these references could go stale...). However this meant that the Future could keep the Storage alive, and thus keep its memory allocated, even after the user was done with it. We fix it here by instead storing a weak ptr to that Storage (well, in fact to the StorageImpl, but it's the same).

Differential Revision: [D29454104](https://our.internmc.facebook.com/intern/diff/D29454104/)

[ghstack-poisoned]
